File: backend/src/sqs/router.js
import { 
  executeNoteGeneration, 
  executeTheoryGeneration, 
  executePromptOptimization 
} from './worker.js';

import { 
  handleNoteGenerationFailure, 
  handleTheoryGenerationFailure,
} from './failures.js';

// Using your exact explicit model import path
import TempPromptJob from '../models/temp.model.js';

/**
 * Parses and routes incoming background SQS messages to their 
 * respective heavy AI processing execution pathways.
 * * @param {Object} message - Raw message payload unmarshalled from the SQS event string
 */
export const routeIncomingAiJob = async (message) => {
  const jobType = message.type || 'dsa';

  try {
    if (jobType === 'theory') {
      await executeTheoryGeneration(message);
    } else if (jobType === 'optimize_prompt') {
      await executePromptOptimization(message);
    } else {
      await executeNoteGeneration(message);
    }
  } catch (error) {
    console.error(`[SQS Router Catch] Handling job crash for type ${jobType}: ${error.message}`);

    // Gracefully handle catastrophic pipeline exceptions based on specific transaction domains
    if (jobType === 'optimize_prompt') {
      try {
        const job = await TempPromptJob.findById(message.job_id);
        if (job) {
          job.status = 'failed';
          // Save error string context into instructions tracking space
          job.optimized_instructions = `Prompt Optimization Stream Fault: ${error.message}`;
          await job.save();
        }
      } catch (dbError) {
        console.error(`[SQS Router Fatal] Failed to flag prompt tracker fallback state: ${dbError.message}`);
      }
    } else if (jobType === 'theory') {
      await handleTheoryGenerationFailure(message.theory_id, error.message);
    } else {
      await handleNoteGenerationFailure(message.note_id, error.message);
    }

    // Rethrow error so SQS message acknowledges the transaction failed and increments retry backoffs
    throw error;
  }
};

File: backend/app/sqs/router.py
from pydantic import ValidationError

from .worker import (
    NonRetryableJobError,
    execute_note_generation, 
    execute_theory_generation, 
    execute_prompt_optimization
)
from .handle_failures import (
    handle_note_generation_failure, 
    handle_theory_generation_failure,
    handle_prompt_optimization_failure
)


JOB_REGISTRY = {
    "theory": {
        "executor": execute_theory_generation,
        "id_field": "theory_id",
        "failure_handler": handle_theory_generation_failure,
    },
    "optimize_prompt": {
        "executor": execute_prompt_optimization,
        "id_field": "job_id",
        "failure_handler": handle_prompt_optimization_failure,
    },
    "dsa": {
        "executor": execute_note_generation,
        "id_field": "note_id",
        "failure_handler": handle_note_generation_failure,
    }
}


async def route_incoming_ai_job(message: dict) -> None:
    job_type = message.get("type", "dsa")

    if job_type not in JOB_REGISTRY:
        raise ValueError(f"Unsupported job type: {job_type}")
    
    job_config = JOB_REGISTRY[job_type]
    executor = job_config["executor"]
    failure_handler = job_config["failure_handler"]
    target_id = message.get(job_config["id_field"])

    try:
        await executor(message)
            
    except (ValidationError, KeyError, NonRetryableJobError) as e:
        reason = f"Payload Validation Fault: {e}"
        await failure_handler(target_id, reason)
        return

    except Exception as e:
        reason = f"Background Processing Error: {e}"
        await failure_handler(target_id, reason)
        raise e